sql/inspect: run INSPECT against a single index

spilchen opened the pull request:
Previously, the INSPECT command was a no-op that performed no actual checking. This change implements the first consistency check for INSPECT: index consistency checking, currently supporting a single index at a time.

The consistency checking uses a full outer join between the primary index and secondary index access paths to identify rows that exist in one index but not the other, detecting potential corruption or inconsistency issues.

This new functionality is integrated into the INSPECT job and triggered through SCRUB when the enable_scrub_job setting is enabled. Future work will expose this through direct INSPECT SQL commands.

Note for reviewers:
- Several tests currently generate internal errors in the SQL query engine. These are not logged as issues for now and will be resolved in subsequent work.

Release note: None
Informs #148299

Epic: CRDB-30356

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/150918)
<!-- Reviewable:end -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

@spilchen Great work!

> @fqazi reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-:-OZ9b1wbBuPkVfPyxXY4:b-bhrgau)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZ9ale-7WrY8z59sF2X:-OZ9ale-7WrY8z59sF2Y:b-kdc23o) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L62)):*
> ```Go
> // Start implements the inspectCheck interface.
> func (c *indexConsistencyCheck) Start(
> 	ctx context.Context, cfg *execinfra.ServerConfig, span roachpb.Span, workerIndex int,
> ```

Longer term for pause / resume support are we planning on tracking the spans completed in the job?
___
*[`pkg/sql/inspect/index_consistency_check.go` line 330 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZ9Zwb-3ow8HVY8a1gp:-OZ9Zwb-3ow8HVY8a1gq:bd6xol) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L330)):*
> ```Go
> 			// keys in the primary index.
> 			if idx.IsPartial() {
> 				return unimplemented.New(
> ```

Will this cause the entire inspect job to fail or just skip the secondary index with a warning?


<!-- Sent from Reviewable.io -->

bghal reviewed (approved):
> @bghal reviewed 12 of 12 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150918)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rafiss and @spilchen)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZAQmvA-s7QNrGHbauy:-OZAQmvA-s7QNrGHbauz:blky4g7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L324)):*
> ```Go
> 		for _, idx := range c.tableDesc.PublicNonPrimaryIndexes() {
> 			if idx.GetID() != c.indexID {
> 				continue
> ```

nit: since the code below only runs once can it be factored out of the loop?
___
*[`pkg/sql/inspect/index_consistency_check.go` line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZA3l_DCYQs3Tirup7O:-OZA3l_DCYQs3Tirup7P:bw1r91y) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L600)):*
> ```Go
> 		return u.String()
> 	}
> 	return fmt.Sprintf("%s.%s", tableAlias, &u)
> ```

nit: use `.String()` directly?
___
*[`pkg/sql/inspect/index_consistency_check.go` line 647 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZABhuk3EE4FtD0KQyN:-OZABhuk3EE4FtD0KQyO:b-xg374v) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L647)):*
> ```Go
> 
> // buildWhereClause constructs WHERE clauses for the query based on predicate and null filters.
> func buildWhereClause(predicate string, nullFilters []string, includeNullFilter bool) string {
> ```

nit: maybe a string builder? could `includeNullFilter` be implicit?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150918)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @bghal, @fqazi, and @rafiss)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZ9ale-7WrY8z59sF2X:-OZAAx6CDnsxCj59_xYx:b-xlddho) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L62)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Longer term for pause / resume support are we planning on tracking the spans completed in the job?
</blockquote></details>

Yes, this is planned in a follow-on task: https://github.com/cockroachdb/cockroach/issues/148297
___
*[`pkg/sql/inspect/index_consistency_check.go` line 324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZAQmvA-s7QNrGHbauy:-OZAhXhP4fMkzONDUC0z:b-o4yu92) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L324)):*
<details><summary><i>Previously, bghal wrote…</i></summary><blockquote>

nit: since the code below only runs once can it be factored out of the loop?
</blockquote></details>

Do you mean like have a helper that returns the `idx` that matches `indexID` and not have a loop in this function?
___
*[`pkg/sql/inspect/index_consistency_check.go` line 330 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZ9Zwb-3ow8HVY8a1gp:-OZABIYq5GOHGKd5_r1y:b-khoo4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L330)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Will this cause the entire inspect job to fail or just skip the secondary index with a warning?
</blockquote></details>

Good point. Currently, I'm targeting a single index. But when multiple indexes are supported, it does bring up an interesting point. I'm leaning towards keeping this as an error and fail the job. If we just return a warning (or NOTICE?) it may give the wrong impression that everything is fine with the index. So, I feel like an error is the right way to go here.


<!-- Sent from Reviewable.io -->

bghal commented:
*[`pkg/sql/inspect/index_consistency_check.go` line 324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZAQmvA-s7QNrGHbauy:-OZAnKiQ-Cn4ENSh5KGm:bb77ucv) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L324)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Do you mean like have a helper that returns the `idx` that matches `indexID` and not have a loop in this function?
</blockquote></details>

Yep.


<!-- Sent from Reviewable.io -->

bghal commented:
*[`pkg/sql/inspect/index_consistency_check.go` line 324 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZAQmvA-s7QNrGHbauy:-OZAo9RbCsqI3TWNmJGC:bqxsynq) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L324)):*
<details><summary><i>Previously, bghal wrote…</i></summary><blockquote>

Yep.
</blockquote></details>

Actually misread that before. Looks good as is.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150918)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @bghal, @fqazi, and @spilchen)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 330 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZ9Zwb-3ow8HVY8a1gp:-OZBFXYT8gPmoXsPPym5:bf6jpwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L330)):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

Good point. Currently, I'm targeting a single index. But when multiple indexes are supported, it does bring up an interesting point. I'm leaning towards keeping this as an error and fail the job. If we just return a warning (or NOTICE?) it may give the wrong impression that everything is fine with the index. So, I feel like an error is the right way to go here.
</blockquote></details>

another option is to not return the error, but instead log that the partial index (or vector index or inverted index) could not be validated.

not being able to validate any table that has any of these index types feels a bit overly restrictive.

for the partial index case, we could also try to come up with a modified query that validates that the rows are a subset of the primary index. (this could be file as a follow up issue)
___
*[`pkg/sql/inspect/index_consistency_check.go` line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZA3l_DCYQs3Tirup7O:-OZBGwb1-BWjcK9rH7AR:b3df96c) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L600)):*
<details><summary><i>Previously, bghal wrote…</i></summary><blockquote>

nit: use `.String()` directly?
</blockquote></details>

it might be a bit more explicit to use `lexbase.EncodeRestrictedSQLIdent` rather than relying on the `String` behavior to quote this. the TTL job uses that function too.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150918)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @bghal, @fqazi, and @rafiss)

___
*[`pkg/sql/inspect/index_consistency_check.go` line 330 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZ9Zwb-3ow8HVY8a1gp:-OZFF8wb8m4_tiRXhIZa:b-jr103k) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L330)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

another option is to not return the error, but instead log that the partial index (or vector index or inverted index) could not be validated.

not being able to validate any table that has any of these index types feels a bit overly restrictive.

for the partial index case, we could also try to come up with a modified query that validates that the rows are a subset of the primary index. (this could be file as a follow up issue)
</blockquote></details>

Okay, I see your point. It makes more sense to generate a NOTICE and skip the index. The issue is that this code runs during the job execution, which is too late to issue a NOTICE. We need to handle that earlier, at job creation time. So I’ll keep the unimplemented errors here for now. The logic added to `scrub` is only temporary until we have the actual `INSPECT` command. I’d rather not add special handling for unsupported indexes in `scrub`. Instead, I’ll add a TODO to note that unsupported indexes should be skipped when the `INSPECT` job is created.
___
*[`pkg/sql/inspect/index_consistency_check.go` line 600 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZA3l_DCYQs3Tirup7O:-OZF2ewX3KMKOt7CQLZY:b9ld37p) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L600)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it might be a bit more explicit to use `lexbase.EncodeRestrictedSQLIdent` rather than relying on the `String` behavior to quote this. the TTL job uses that function too.
</blockquote></details>

Good idea. I see there are a few other places where I need to use this. I added a test where the table has column names that match reserver words to verify.
___
*[`pkg/sql/inspect/index_consistency_check.go` line 647 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/150918#-OZABhuk3EE4FtD0KQyN:-OZF5Qv09eKznGl3p6OD:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf60fbf438644402dc9b33e9446771c152f65769/pkg/sql/inspect/index_consistency_check.go#L647)):*
<details><summary><i>Previously, bghal wrote…</i></summary><blockquote>

nit: maybe a string builder? could `includeNullFilter` be implicit?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: run INSPECT against a single index

Previously, the INSPECT command was a no-op that performed no actual
checking. This change implements the first consistency check for
INSPECT: index consistency checking, currently supporting a single
index at a time.

The consistency checking uses a full outer join between the primary
index and secondary index access paths to identify rows that exist
in one index but not the other, detecting potential corruption or
inconsistency issues.

This new functionality is integrated into the INSPECT job and
triggered through SCRUB when the enable_scrub_job setting is enabled.
Future work will expose this through direct INSPECT SQL commands.

Note for reviewers:
- Some tests currently generate internal errors in the SQL query
  engine. These are not logged as INSPECT issues for now and will be
  resolved in subsequent work.
- I include a crdb_internal.void_func() predicate in the query to
  prevent the optimizer from optimizing the join away. There were
  a couple of other options (add barrier option to CTE, materialize
  the CTE inline), but this was the preferred method.

Release note: None

Epic: CRDB-30356

spilchen commented:
TFTRs!

bors r+

spilchen commented:
bors retry
